Fix issue for debug containers when using custom Docker registry

javaducky opened the pull request:
**Subject**
Fixes bug where override of Docker registry was not being applied to debug containers (#3851)

**Problem**
Overrides for Docker registry are not being applied to debug containers and provide no means to correct the image.

**Solution**
This update expands the `data.proxy` configuration section within the Linkerd `ConfigMap` to maintain the overridden image name for debug containers at _install_-time similar to handling of the `proxy` and `proxyInit` images.

This change also enables the further override option of the registry for debug containers at _inject_-time given utilization of the `--registry` CLI option.

**Validation**
Several new unit tests have been created to confirm functionality.  In addition, the following workflows were run through:

### Standard Workflow with Custom Registry
This workflow installs Linkerd control plane based upon a custom registry, then injecting the debug sidecar into a service.

* Start with a k8s instance having no Linkerd installation
* Build all images locally using `bin/docker-build`
* Create custom tags (using same version) for generated images, e.g. `docker tag gcr.io/linkerd-io/debug:git-a4ebecb6 javaducky.com/linkerd-io/debug:git-a4ebecb6`
* Install Linkerd with registry override `bin/linkerd install --registry=javaducky.com/linkerd-io | kubectl apply -f -`
* Once Linkerd has been fully initialized, you should be able to confirm that the `linkerd-config` ConfigMap now contains the debug image name, pull policy, and version within the `data.proxy` section
* Request injection of the debug image into an available container.  I used the Emojivoto voting service as described in https://linkerd.io/2/tasks/using-the-debug-container/ as `kubectl -n emojivoto get deploy/voting -o yaml | bin/linkerd inject --enable-debug-sidecar - | kubectl apply -f -`
* Once the deployment creates a new pod for the service, inspection should show that the container now includes the "linkerd-debug" container name based on the applicable override image seen previously within the ConfigMap
* Debugging can also be verified by viewing debug container logs as `kubectl -n emojivoto logs deploy/voting linkerd-debug -f`
* Modifying the `config.linkerd.io/enable-debug-sidecar` annotation, setting to “false”, should show that the pod will be recreated no longer running the debug container.

### Overriding the Custom Registry Override at Injection
This builds upon the “Standard Workflow with Custom Registry” by overriding the Docker registry utilized for the debug container at the time of injection.

* “Clean” the Emojivoto voting service by removing any Linkerd annotations from the deployment
* Request injection similar to before, except provide the `--registry` option as in `kubectl -n emojivoto get deploy/voting -o yaml | bin/linkerd inject --enable-debug-sidecar --registry=gcr.io/linkerd-io - | kubectl apply -f -`
* Inspection of the deployment config should now show the override annotation for `config.linkerd.io/debug-image` having the debug container from the new registry.  Viewing the running pod should show that the `linkerd-debug` container was injected and running the correct image.  Of note, the proxy and proxy-init images are still running the “original” override images.
* As before, modifying the `config.linkerd.io/enable-debug-sidecar` annotation setting to “false”, should show that the pod will be recreated no longer running the debug container.

### Standard Workflow with Default Registry
This workflow is the typical workflow which utilizes the standard Linkerd image registry.

* Uninstall the Linkerd control plane using `bin/linkerd install --ignore-cluster | kubectl delete -f -` as described at https://linkerd.io/2/tasks/uninstall/
* Clean the Emojivoto environment using `curl -sL https://run.linkerd.io/emojivoto.yml | kubectl delete -f -` then reinstall using `curl -sL https://run.linkerd.io/emojivoto.yml | kubectl apply -f -`
* Perform standard Linkerd installation as `bin/linkerd install | kubectl apply -f -`
* Once Linkerd has been fully initialized, you should be able to confirm that the `linkerd-config` ConfigMap references the default debug image of `gcr.io/linkerd-io/debug` within the `data.proxy` section
* Request injection of the debug image into an available container as `kubectl -n emojivoto get deploy/voting -o yaml | bin/linkerd inject --enable-debug-sidecar - | kubectl apply -f -`
* Debugging can also be verified by viewing debug container logs as `kubectl -n emojivoto logs deploy/voting linkerd-debug -f`
* Modifying the `config.linkerd.io/enable-debug-sidecar` annotation, setting to “false”, should show that the pod will be recreated no longer running the debug container.

### Overriding the Default Registry at Injection
This workflow builds upon the “Standard Workflow with Default Registry” by overriding the Docker registry utilized for the debug container at the time of injection.

* “Clean” the Emojivoto voting service by removing any Linkerd annotations from the deployment
* Request injection similar to before, except provide the `--registry` option as in `kubectl -n emojivoto get deploy/voting -o yaml | bin/linkerd inject --enable-debug-sidecar --registry=javaducky.com/linkerd-io - | kubectl apply -f -`
* Inspection of the deployment config should now show the override annotation for `config.linkerd.io/debug-image` having the debug container from the new registry.  Viewing the running pod should show that the `linkerd-debug` container was injected and running the correct image.  Of note, the proxy and proxy-init images are still running the “original” override images.
* As before, modifying the `config.linkerd.io/enable-debug-sidecar` annotation setting to “false”, should show that the pod will be recreated no longer running the debug container.

Fixes issue #3851 

Signed-off-by: Paul Balogh javaducky@gmail.com

alpeb commented:
Thanks @javaducky, looks pretty exhaustive! :tada: 
I've just started to review this. When running the full integration suite I found there's an error when trying to upgrade from the latest stable (2.6.1) version. In `cli/cmd/inject.go` when you try doing
```
configs.Debug.DebugImage.PullPolicy = options.imagePullPolicy
```
It causes a panic, presumably because `configs.Debug` is empty in 2.6.1, so you need to account for that case.
FYI the deep integration tests aren't run automatically in github for PRs, but you can run full suite in your machine as explained [here](https://github.com/linkerd/linkerd2/blob/master/TEST.md#integration-tests).

javaducky commented:
Awesome, thank you!  I'll get that fixed and make sure to run those tests locally.  Lots of angles with this ;)

javaducky commented:
@alpeb I've corrected the `upgrade` issue and have confirmed that the integration tests are able to fully pass.  I apologize for not having run the full suite previously.

alpeb commented:
Thanks @javaducky . Do you mind merging against master before we resume reviewing?

javaducky commented:
@alpeb, this should be good again given the rebase to latest master.

javaducky commented:
@adleong thanks, I'm definitely having fun with this project!  As for your suggestion, including the debug configuration options in the `data.proxy` section was one of the original thoughts that had been discussed in Slack (conversations start around [December 27](https://linkerd.slack.com/archives/CMS5ZD60Y/p1577486748046600)).  I only received feedback on that from @alpeb given the holidays, so I went with it.

Having the separation from the proxy felt a bit cleaner to me, but I'm totally fine with including it in the `data.proxy` section if that's preferred.  Not sure if any others might be interested in taking a quick look at the [options I had posted](https://linkerd.slack.com/archives/CMS5ZD60Y/p1577487521050400) originally.  Option 3 is what is currently implemented, Option 2 seems to match what you're suggesting.

As I mentioned, I'm fine either way and it's not a huge pain to change if one option is deemed preferable over another.  Other than including @alpeb again, I'm unsure of who else should be asked to weigh their opinion.  @grampelberg I'm guessing?

Please let know!

adleong commented:
I chatted briefly with @alpeb and we think it makes sense to put this in the proxy config (i.e. option 2).

javaducky commented:
Cool...not a problem!  I should have those changes turned around pretty quick.

javaducky commented:
@adleong and @alpeb .  I've now updated the code to no longer use the custom `data.debug` configuration section any longer.  Now all of the debug container information will be included in the `data.proxy` section.  I've rebased the changes against the latest master.

javaducky commented:
Updated "Validation" section of PR given most recent usage change.

javaducky commented on charts/linkerd2/README.md:
```diff
@@ -97,6 +97,9 @@ The following table lists the configurable parameters of the Linkerd2 chart and
 | `dashboard.replicas`                  | Number of replicas of dashboard                                                                                                                                                       | `1`                                  |
+| `debugContainer.image.name`           | Docker image for the debug container                                                                                                                                                  | `gcr.io/linkerd-io/proxy`            |
+| `debugContainer.image.pullPolicy`     | Pull policy for the debug container Docker image                                                                                                                                      | `IfNotPresent`                       |
+| `debugContainer.image.version`        | Tag for the debug container Docker image                                                                                                                                              | `stable-2.5.0`                       |
```
Sure thing!  I meant to ping you to ask about those when I saw them.  I think they're like this in the latest edge release.

javaducky replied:
@adleong I pulled in your merged updates and it appears you missed the `global.proxyInit.image.version` which is actually v1.3.0 now.  Shall I apply the same "latest version" treatment to this as well?

javaducky replied:
Actually, prometheus has been updated by @Pothulapati and should be v2.15.2.  Just let me know if you'd rather those be on another issue/PR.

adleong replied:
Good call!

Pothulapati replied:
@javaducky I'm actually doing a clean up of the read-me with adding more flags, and sorting them alphabetically. I can take this up too.

javaducky commented on cli/cmd/inject.go:
```diff
@@ -479,3 +490,12 @@ func parsePortRanges(portRanges []*cfg.PortRange) string {
 	return strings.TrimSuffix(str, ",")
 }
+
+func getFlagValue(flags []*cfg.Install_Flag, name string) string {
```
The reason I included this was so that if the `install` was done with a custom registry, an override at `inject`-time would correctly override that custom registry rather than expecting the default `gcr.io/linkerd-io` for the override.  I believe this should be applied to your prior PR for the proxy and proxy-init as well.  I'll have to look at it again more closely this evening, but I believe that to be the case if I'm not mistaken.

javaducky replied:
@adleong I've confirmed that we do need this and was able to verify with additional unit tests.  Here's the process I'm guarding against:

Install using `linkerd install --registry=install.registry` results in ConfigMap settings like:
```
data:
  proxy: {"imageName":"install.registry/proxy",...}
  install: {"flags":[{"name":"registry", "value":"install.registry"}],...}
```

Inject using `linkerd inject --registry=inject.registry` will still only inject `install.registry/proxy`.

The injected image is not overridden because the current `registryOverride()` method in `install.go` only replaces the _default_ registry, i.e. `gcr.io/linkerd-io`.  In our case, we want to have the new registry override `install.registry` which we only know by looking at the installation flags.

The `registryOverride()` method in `install.go` is fine as it is, but should not be used by `inject.go`.

adleong replied:
Ah, nice catch then.  It seems to me that this registry replacement logic has always been quite brittle.  Could we simplify things by just using string splitting or a regex to replace the registry portion of the string, without needing to know the previous value?

javaducky replied:
We could for the debug image since it has no other way to override the image name other than the registry portion (i.e. we don't allow a CLI option of `--debug-image`) and therefore we'd replace anything prior to "/debug."  As for proxy and proxy-init, we don't really have that luxury as during install, someone could use `my.custom.registry/myproxy`.  I _think_ this is probably the closest we can get offhand, though still not 100% infallible due to the flexibility allowed with the CLI.

adleong replied:
Hmmm... there must be something I'm missing here, because I still don't see why we need to know the previous registry value.  This is the flow as I understand it:

Suppose Linkerd is installed with `--registry my-install-registry.io/linkerd-io` and `--proxy-image gcr.io/my-install-image`.  I would expect that the proxy-image flag would take priority and the resulting proxy image name of `gcr.io/my-install-image` would be set into the Linkerd configmap.  Since there is no flag for setting the debug image name, `my-install-registry.io/linkerd-io/debug` is set as the debug image name in the configmap.

Then suppose that inject is run with `--registry my-inject-registry.io/linkerd-io` and `--proxy-image gcr.io/my-inject-image`.  For the proxy, again the proxy image flag takes priority and we use `gcr.io/my-inject-image` as the proxy image name instead of the value in the configmap.  For the debug container, we take the value from the configmap (`my-install-registry.io/linkerd-io/debug`) and replace the registry part (everything up to the last slash) with the registry override from the inject flag: `my-inject-registry.io/linkerd-io/debug`.

Am I missing something about the way this works?

javaducky replied:
@adleong Ok, now I see!  Sorry...wasn't grokking that the last slash truly indicates the end of the registry name; I guess I was thinking of a more convoluted registry and image name where a slash "/" could be part of the image name which cannot be.  

I'll make that change...definitely simplifies the replacement. 👍

javaducky commented on cli/cmd/inject.go:
```diff
@@ -392,9 +392,20 @@ func (options *proxyConfigOptions) overrideConfigs(configs *cfg.All, overrideAnn
 		overrideAnnotations[k8s.ProxyImagePullPolicyAnnotation] = options.imagePullPolicy
 	}
 
+	if options.dockerRegistry != "" {
```
Ah, yes.  I forgot to look at that again after rebasing with your previous changes.

javaducky commented on charts/linkerd2/README.md:
```diff
@@ -97,6 +97,9 @@ The following table lists the configurable parameters of the Linkerd2 chart and
 | `controllerUID`                       | User ID for the control plane components                                                                                                                                              | `2103`                               |
 | `identity.issuer.clockSkewAllowance`  | Amount of time to allow for clock skew within a Linkerd cluster                                                                                                                       | `20s`                                |
 | `dashboard.replicas`                  | Number of replicas of dashboard                                                                                                                                                       | `1`                                  |
+| `debugContainer.image.name`           | Docker image for the debug container                                                                                                                                                  | `gcr.io/linkerd-io/proxy`            |
```
Ooops.  Totally wasn't thinking there.  Thanks for catching that!

javaducky commented on cli/cmd/inject.go:
```diff
@@ -367,10 +367,18 @@ func (options *proxyConfigOptions) overrideConfigs(configs *cfg.All, overrideAnn
-		overrideAnnotations[k8s.ProxyInitImageAnnotation] = configs.Proxy.ProxyInitImage.ImageName
+		currentProxyImage := configs.Proxy.ProxyImage.ImageName
+		currentProxyInitImage := configs.Proxy.ProxyInitImage.ImageName
+		currentDebugImage := configs.Proxy.DebugImage.ImageName
```
Yeah, I should've thought about that.  Usually try to make sure to use the `Get...` on non-mutating calls.  Will update.

javaducky commented:
@alpeb fyi, the tests are currently failing; presumably until https://github.com/linkerd/linkerd2/pull/3923/files is merged into master and rebased into this branch.  I'll let you know once it should be safe to continue review.

javaducky commented on cli/cmd/inject.go:
```diff
@@ -367,10 +367,13 @@ func (options *proxyConfigOptions) overrideConfigs(configs *cfg.All, overrideAnn
+		}
+		overrideAnnotations[k8s.ProxyImageAnnotation] = overwriteRegistry(configs.GetProxy().GetProxyImage().GetImageName(), options.dockerRegistry)
+		overrideAnnotations[k8s.ProxyInitImageAnnotation] = overwriteRegistry(configs.GetProxy().GetProxyInitImage().GetImageName(), options.dockerRegistry)
+		overrideAnnotations[k8s.DebugImageAnnotation] = overwriteRegistry(debugImage, options.dockerRegistry)
```
@adleong Here is the "simplified" usage to overwrite whatever registry was used without looking up the registry flag.

I started down the path of even replacing the `registryOverride` method in `root.go` with this implementation, but unit tests were broken around the `install-cni-plugin.go` so I backed off.  What do you think?  Should the `registryOverride` implementation be changed and fix the test accordingly?  I _feel_ like that's really what we would want; not just overriding if the default registry was being used, but the method description was very specific, so possibly very much intended.

adleong replied:
I'd say it's definitely confusing that there are both `registryOverride` and `overwriteRegistry` methods that do subtly different things.  That said, I don't want this PR to creep in scope forever.  I think my ideal solution would be to go ahead with this and then do a followup PR in which we can take a hard look at `registryOverride` and figure out why it's like that and find a way to unify or differentiate the two functions.

javaducky replied:
Agreed.

adleong commented on cli/cmd/inject.go:
```diff
@@ -479,3 +483,19 @@ func parsePortRanges(portRanges []*cfg.PortRange) string {
+	if !strings.HasSuffix(updated, "/") {
+		updated += "/"
+	}
+	i := len(image) - 1
```
Could you use `strings.LastIndex` here?

javaducky replied:
Yeah...didn't notice that method.  I'll update...I have a merge conflict to contend with anyway. 😄

javaducky commented:
@adleong @alpeb this should be ready for you again. 🤞

javaducky commented:
Yeah, I forgot that I left that in after removing the lookup of install flags.
